activerecord/lib/active_record/railtie.rb

```diff
@@ -162,10 +162,13 @@ class Railtie < Rails::Railtie # :nodoc:
                 warn "Failed to validate the schema cache because of #{error.class}: #{error.message}"
                 nil
               end
-              next if current_version.nil?
 
-              if cache.schema_version != current_version
+              if current_version.nil?
+                connection_pool.schema_reflection.clear!
+                next
+              elsif cache.schema_version != current_version
                 warn "Ignoring #{filename} because it has expired. The current schema version is #{current_version}, but the one in the schema cache file is #{cache.schema_version}."
+                connection_pool.schema_reflection.clear!
```

Comment:
This sounds reasonable but only semi-related... assuming it is separate (the instigating change?), might be better split out, it only because of how confusing it is that this is also about a different version check? 😅

Reply:
Moved to https://github.com/rails/rails/pull/49415